src/connection_manager.rs

```
@@ -382,31 +402,49 @@ impl ConnectionManager {
     // nodes we should establish connections with
     async fn bootstrap_and_handshake(&mut self) -> Result<Vec<SocketAddr>, Error> {
         trace!("Bootstrapping with contacts...");
-        let (endpoint, conn, _incoming_messages) = self.qp2p.bootstrap().await?;
+        let (endpoint, conn, mut incoming_messages) = self.qp2p.bootstrap().await?;
         self.endpoint = Arc::new(Mutex::new(endpoint));
 
         trace!("Sending handshake request to bootstrapped node...");
         let public_key = self.keypair.public_key();
         let handshake = HandshakeRequest::Bootstrap(public_key);
         let msg = Bytes::from(serialize(&handshake)?);
-        let mut streams = conn.send_bi(msg).await?;
-        let response = streams.1.next().await?;
-
-        match deserialize(&response) {
-            Ok(HandshakeResponse::Rebootstrap(_elders)) => {
-                trace!("HandshakeResponse::Rebootstrap, trying again");
-                // TODO: initialise `hard_coded_contacts` with received `elders`.
-                unimplemented!();
-            }
-            Ok(HandshakeResponse::Join(elders)) => {
-                trace!("HandshakeResponse::Join Elders: ({:?})", elders);
-
-                // Obtain the addresses of the Elders
-                let elders_addrs = elders.into_iter().map(|(_xor_name, ci)| ci).collect();
-                Ok(elders_addrs)
+        match conn.send_bi(msg).await {
+            Ok(_) => {
+                if let Some(message) = incoming_messages.next().await {
+                    match message {
+                        Qp2pMessage::BiStream { bytes, .. }
+                        | Qp2pMessage::UniStream { bytes, .. } => {
```

Can nodes be connected to the client via `UniStream`? Think we should return an error here.

---

I'm not sure, but once they have the connection created I think both sides can create any type of streams?

---

We're getting uni messages back at the mo when we do `routing.send_to_client` at nodes. What qp2p is doing under the hood I'm not sure. Sending a uni back on the original bistream i guess?